### ISSUE TITLE
Notifications using Javascript Notification API

### ISSUE THREAD

**Issue by hilariocoelho** [go-gitea/gitea#10804]
## Description
Current Gitea's notification system is kinda useless in my opinion. I get a lot of spam of notifications (new issues, closed issues, new issues comments, merged PRs (in a micro service project this becomes an huge spam of notifications), etc). When I actually see my notifications, there are usually dozens of them so I just click to mark them all as seen without actually seeing them.

Gitea should use [Notifications API](https://developer.mozilla.org/en-US/docs/Web/API/Notifications_API/Using_the_Notifications_API) so I can see System Notifications while I use my IDE or something.

Gitea should also support the user to setup which events he wants to get notifications from on the settings page, ie I want to get notifictions only for new issues, comments on issues and comments on PRs.

Maybe gitea should use something like [Onesignal](https://onesignal.com/webpush) so it can send push notifications even if the user doesn't have any gitea's tab opened.

This issue is an improvement of #9327

**Comment by SpaWn2KiLl** [comment]
I think this is also an improvement of #10198 
@6543

**Comment by jamesorlakin** [comment]
Remember a big goal of Gitea is being self-hosted, so using external options like Onesignal causes a form of vendor lock-in and a dependency on a user setting this service up correctly.

The [Web Push API](https://developer.mozilla.org/en-US/docs/Web/API/Push_API) exists for this purpose, where the browser vendor gives you a URL to post messages to. This will mean Gitea needs to keep a track of these URLs when a user has opted in and post notifications to it as they are generated. The service worker can receive these and use the normal notifications API you linked above. This avoids the need to poll for messages and should work when a tab isn't open. I have questions over how we would know to remove older entries over time however. [This guide](https://medium.com/izettle-engineering/beginners-guide-to-web-push-notifications-using-service-workers-cb3474a17679) proves quite useful in giving an overview.

We would also need to come up with a 'notification reason', if that doesn't exist already. "Why am I being notified by this PR?": _x commented 4 hours ago..._

**Comment by SpaWn2KiLl** [comment]
Didn't know about the Web Push API! Thanks for the information!

I did some more research and found another good article that explains better the Subscription ID and the server side integration [here](https://flaviocopes.com/push-api/). There is also an easy to use Golang library called [webpush-go](https://github.com/SherClockHolmes/webpush-go).

**Comment by silverwind** [comment]
Agree, Notifications should be more configurable e.g. more than just watch/unwatch.

I don't think Push/Web Notifications are desirable. These APIs are finicky and they would likely need web sockets which can be hard to configure and they can also be a source of security issues.

**Comment by mohe2015** [comment]
I don't think they need websockets. As far as I know you simply send the subscription json to the server using a POST request and it can then use it to send push messages to you. See e.g. https://developers.google.com/web/fundamentals/push-notifications

**Comment by jamesorlakin** [comment]
I would disagree - push notifications would potentially supplant notifications by webhook in some way. In our team we have a Slack channel dedicated for notifications from Gitea and Drone. These are a mix of projects and therefore there's a lot of 'noise'. Since Slack doesn't have custom notifications for channels it often gets muted and then new PRs, etc aren't necessarily picked up by people.

Having a push notification from the browser direct would negate the need for an external system for this workflow. No websockets needed! We already have a service worker too. The APIs don't seem great no, but it's definitely doable I think.

I might have a play over the weekend. 😃

**Comment by jamesorlakin** [comment]
Update: I've had a bit of a play and am actually making some progress :tada:

The big bit of what's left to do is storing the web subscriptions in a database (and therefore making the model, tiny API, etc). Also localisation and linking direct to the comment would be nice.

As for giving more context to the notification I'm not too sure - This is a separate issue and PR I feel, as overhauling notifications is a big task in its own right. I can see some similarities in logic to activities and webhooks a little though.

![Gitea notifications](https://user-images.githubusercontent.com/7294642/77836370-8a018f80-714d-11ea-9754-68e3f1667b90.gif)

![image](https://user-images.githubusercontent.com/7294642/77836410-cfbe5800-714d-11ea-9a8d-bee4a21fdfe6.png)

**Comment by jamesorlakin** [comment]
I'm also considering the possible future enhancements having a live server to client push technology will bring. By getting the service worker to trigger events on open webpages we _could_:
- Automatically increment the bell icon
- On open issues and PRs, pop a message up saying there's new content.

**Comment by SpaWn2KiLl** [comment]
It is looking good! 🙂

Are you using Notifications API? Is that creating system notifications or is it a custom notification box?

**Comment by jamesorlakin** [comment]
It's a mixture of the Web Push API to get JSON messages to the service worker in the background and then the normal Notfications API to display it. It's the default look for Firefox - I don't think there's much customisability though.

**Comment by 6543** [comment]
do  #10961 close this?

**Comment by catdevnull** [comment]
No, this is not fixed by  #10961 as it doesn't do webpush notifications. #10884 does.

**Comment by karmanyaahm** [comment]
Hi,
If I were to submit a PR implementing just the backend/API part of this, would it be accepted?

Setting up the backend would let Android apps such as GitNex receive WebPush notifications using the UnifiedPush standard (they're compatible) (https://codeberg.org/gitnex/GitNex/issues/1086). Additionally, someone who is better at JS than me could implement browser notifications in the future.

This was originally suggested by @6543 in https://github.com/UnifiedPush/wishlist/issues/23

**Comment by MVoikolesco** [comment]
Hi! I'd like to know if you've implemented the backend/API part to support UnifiedPush.

If you're still interested in this functionality, I'd be happy to contribute to the frontend (or review it if it's already completed).

Thank you!
